# TACCSSHClient Class

The TACCSSHClient class is used to establish connections with TACC systems, using a user's login and 2FA code.

The class then provides methods for (1)basic file operations, and for (2) execution of arbitrary commands.
We will go over these two main areas of functionality in the tutorial

## Initialization

To initialize a `TACCSSHClient` class object, you will to pass in the system you want to connect to. Currently supported systems include:

- stampede2 - [Stampede2](https://portal.tacc.utexas.edu/user-guides/stampede)
- ls6 - [Lonestar6](https://portal.tacc.utexas.edu/user-guides/lonestar6)
- frontera - [Frontera](https://frontera-portal.tacc.utexas.edu/user-guide/)
- maverick2 - [Maverick2](https://portal.tacc.utexas.edu/user-guides/maverick2)


In [6]:
from taccjm.TACCSSHClient import TACCSSHClient

tsc = TACCSSHClient("stampede2", "clos21")

Password: ········
TACC Token Code: 322732


> /home/jovyan/work/repos/taccjm/src/taccjm/TACCSSHClient.py(218)process_command()
    216                        "status": command_config['status']}
    217         pdb.set_trace()
--> 218         if command_config["channel"].exit_status_ready() or wait:
    219             command_config['history'].append(prev_status)
    220             command_config["rc"] = command_config["channel"].recv_exit_status()



ipdb>  c


> /home/jovyan/work/repos/taccjm/src/taccjm/TACCSSHClient.py(218)process_command()
    216                        "status": command_config['status']}
    217         pdb.set_trace()
--> 218         if command_config["channel"].exit_status_ready() or wait:
    219             command_config['history'].append(prev_status)
    220             command_config["rc"] = command_config["channel"].recv_exit_status()



ipdb>  c


> /home/jovyan/work/repos/taccjm/src/taccjm/TACCSSHClient.py(218)process_command()
    216                        "status": command_config['status']}
    217         pdb.set_trace()
--> 218         if command_config["channel"].exit_status_ready() or wait:
    219             command_config['history'].append(prev_status)
    220             command_config["rc"] = command_config["channel"].recv_exit_status()



ipdb>  c


In [9]:
tsc.execute_command('mamba env list', wait=True)

> /home/jovyan/work/repos/taccjm/src/taccjm/TACCSSHClient.py(218)process_command()
    216                        "status": command_config['status']}
    217         pdb.set_trace()
--> 218         if command_config["channel"].exit_status_ready() or wait:
    219             command_config['history'].append(prev_status)
    220             command_config["rc"] = command_config["channel"].recv_exit_status()



ipdb>  command_config


{'id': 4, 'cmd': 'mamba env list', 'ts': datetime.datetime(2023, 2, 9, 18, 16, 29, 250062), 'status': 'STARTED', 'stdout': '', 'stderr': '', 'history': [], 'channel': <paramiko.Channel 4 (closed) -> <paramiko.Transport at 0x35d3df30 (cipher aes128-ctr, 128 bits) (active; 0 open channel(s))>>, 'rt': None}


ipdb>  n


> /home/jovyan/work/repos/taccjm/src/taccjm/TACCSSHClient.py(219)process_command()
    217         pdb.set_trace()
    218         if command_config["channel"].exit_status_ready() or wait:
--> 219             command_config['history'].append(prev_status)
    220             command_config["rc"] = command_config["channel"].recv_exit_status()
    221             ts = datetime.now()



ipdb>  n


> /home/jovyan/work/repos/taccjm/src/taccjm/TACCSSHClient.py(220)process_command()
    218         if command_config["channel"].exit_status_ready() or wait:
    219             command_config['history'].append(prev_status)
--> 220             command_config["rc"] = command_config["channel"].recv_exit_status()
    221             ts = datetime.now()
    222             command_config["stdout"] += (



ipdb>  


> /home/jovyan/work/repos/taccjm/src/taccjm/TACCSSHClient.py(221)process_command()
    219             command_config['history'].append(prev_status)
    220             command_config["rc"] = command_config["channel"].recv_exit_status()
--> 221             ts = datetime.now()
    222             command_config["stdout"] += (
    223                 command_config["channel"].recv(max_nbytes).decode("utf-8")



ipdb>  command_config


{'id': 4, 'cmd': 'mamba env list', 'ts': datetime.datetime(2023, 2, 9, 18, 16, 29, 250062), 'status': 'STARTED', 'stdout': '', 'stderr': '', 'history': [{'ts': datetime.datetime(2023, 2, 9, 18, 16, 29, 250062), 'status': 'STARTED'}], 'channel': <paramiko.Channel 4 (closed) -> <paramiko.Transport at 0x35d3df30 (cipher aes128-ctr, 128 bits) (active; 0 open channel(s))>>, 'rt': None, 'rc': 127}


ipdb>  c


2023-02-09 18:16:58,449 | taccjm.TACCSSHClient | ERROR | Non-zero return code.
clos21@stampede2.tacc.utexas.edu$ mamba env list
rc     : 127
stdout : 
stderr : bash: mamba: command not found



SSHCommandError: Non-zero return code.
clos21@stampede2.tacc.utexas.edu$ mamba env list
rc     : 127
stdout : 
stderr : bash: mamba: command not found


## Command Operations

### `execute_command()`

The TACCSSHClient offers the `execute_command()` function to execute arbitrary commands as if one were in an ssh session interactively. 
The commands are executed on a separately opened channel, spawned from the original paramiko transport object used to establish the main ssh connection when the class was initialized.

In [2]:
print(tsc.execute_command("showq -u clos21")["stdout"])

Using system default configuration file 

SUMMARY OF JOBS FOR USER: <clos21>

ACTIVE JOBS--------------------
JOBID     JOBNAME    USERNAME      STATE   NODES REMAINING STARTTIME

WAITING JOBS------------------------
JOBID     JOBNAME    USERNAME      STATE   NODES WCLIMIT   QUEUETIME

Total Jobs: 0     Active Jobs: 0     Idle Jobs: 0     Blocked Jobs: 0   



### Command Errors

If the command fails, an error will be thrown:

In [3]:
tsc.execute_command("showr")["stderr"]

{"asctime": "2023-02-07 17:36:29,795", "name": "taccjm.TACCSSHClient", "levelname": "ERROR", "message": "Non-zero return code.\nclos21@ls6.tacc.utexas.edu$ showr\nrc     : 127\nstdout : \nstderr : bash: showr: command not found\n"}


SSHCommandError: Non-zero return code.
clos21@ls6.tacc.utexas.edu$ showr
rc     : 127
stdout : 
stderr : bash: showr: command not found


As we can see, we get a good error clearly indicating:

- system on which command was run (Lonestar6)
- User running command (clos21)
- Command being run, after the "$" sign, (showr)
- The return code (127), stdout, and stderr

So we can clearly see here that the command failed because `showr` is not a valid command. We probably meant `showq`

We can also supress the error by setting `error=False`.
This allows use to check the error code/error message and handle it as necessary.

In [4]:
tsc.execute_command("showr", error=False)

{'id': 4,
 'cmd': 'showr',
 'ts': datetime.datetime(2023, 2, 7, 17, 36, 31, 439860),
 'status': 'FAILED',
 'stdout': '',
 'stderr': 'bash: showr: command not found\n',
 'history': [{'ts': datetime.datetime(2023, 2, 7, 17, 36, 30, 288107),
   'status': 'STARTED'}],
 'rt': 1,
 'rc': 127}

### `process_command()` - Asynchronous commands

Commands can also be run in the background, for example if performing a long copy or monitoring some job. 
This is done by set the `wait=False` parameter, and then using the `process_command()` method to check on the status of the command using the command's ID.

In [5]:
cmnd = tsc.execute_command("echo START; sleep 10; echo START", wait=False)
cmnd

{'id': 5,
 'cmd': 'echo START; sleep 10; echo START',
 'ts': datetime.datetime(2023, 2, 7, 17, 36, 31, 571251),
 'status': 'STARTED',
 'stdout': '',
 'stderr': '',
 'history': [],
 'channel': <paramiko.Channel 4 (open) window=2097152 -> <paramiko.Transport at 0x3d257790 (cipher aes128-ctr, 128 bits) (active; 1 open channel(s))>>,
 'rt': None}

In [6]:
cmnd = tsc.process_command(cmnd["id"])
cmnd

{'id': 5,
 'cmd': 'echo START; sleep 10; echo START',
 'ts': datetime.datetime(2023, 2, 7, 17, 36, 31, 712024),
 'status': 'RUNNING',
 'stdout': '',
 'stderr': '',
 'history': [{'ts': datetime.datetime(2023, 2, 7, 17, 36, 31, 571251),
   'status': 'STARTED'}],
 'channel': <paramiko.Channel 4 (open) window=2097152 -> <paramiko.Transport at 0x3d257790 (cipher aes128-ctr, 128 bits) (active; 1 open channel(s))>>,
 'rt': None}

Note the command's history of status's is stored.
This history is compiled as the process is polled, with duplicate status's removed.
Finally note how the `rt` field gives us a rough runtime for the command.
Once again if the process was run in the background, this runtime is according to how the process was polled, not necessarily how long the command took to run.

### Buffering stdout from commands

You can pass an `nbytes: int=None` parameter to the `process_command()` method to read the next `nbytes` available in the stdout buffer of an actively running command:

In [7]:
import time

cmnd = tsc.execute_command("echo START; sleep 7; echo 1; sleep 6; echo 2", wait=False)

t = 0
while cmnd["status"] not in ["COMPLETE", "FAILED"]:
    cmnd = tsc.process_command(cmnd["id"], wait=False, nbytes=5)
    print(f"stdout at time {t}: {cmnd['stdout']}")
    t += 5
    time.sleep(5)

stdout at time 0: START
stdout at time 5: START

stdout at time 10: START
1

stdout at time 15: START
1
2



Note how the output is appended to the 'stdout' field of the command config as it becomes available, and only `nbytes` are read on any given call.

## File Operations

The TACCSHClient class uses its underlying ssh connection to open sftp connections and provide programmatic access to files on TACC systems.
The following operations are offered to interact with files on TACC systems using the TACCSSHClient Class:

- list_files - Get file and directory info (file stats)
- upload - Upload a file/directory to TACC system via sftp.
- downlaod - Download a file/directory from TACC system via sftp.
- read - Read file contents (text or json) directly from TACC system via sftp.
- write - Write data (string or dictioanry) directly from TACC system via sftp.

Note:

    1. All remote paths (paths on TACC systems) must use unix path separators '/'.
    2. File operations currently aren't "safe." 
       That is file uploads/downloads and reads/writes will overwrite contents on remote/local systems.
    3. If absolute paths aren't specified, paths are relative to a user's $HOME diretory on TACC systems, and relative to the cwd for the local paths.

### `list_files()`

The `list_files()` method provides basic file stat info, much like an `ls` command would in a shell session.

In [2]:
scratch_dir = tsc.list_files("")
scratch_dir

[{'filename': '/scratch/06307/clos21/',
  'st_atime': 1634240870,
  'st_gid': 800588,
  'st_mode': 16832,
  'st_mtime': 1675831233,
  'st_size': 19,
  'st_uid': 856065,
  'ls_str': b'drwx------   1 856065   800588         19 07 Feb 22:40 ?'}]

In [3]:
files = tsc.list_files("", recurse=True)
len(files)

19

In [4]:
print(
    "\n".join(
        [
            x["ls_str"].decode("utf-8")
            for x in files
            if not x["filename"].startswith(".")
        ]
    )
)

drwx------   1 856065   800588          4 19 Aug 05:32 11
drwx------   1 856065   800588          4 03 Jun 2022  taccjm
drwx------   1 856065   800588          4 08 Sep 22:48 test-taccjm
drwx------   1 856065   800588          1 02 Feb 12:00 ch-sim-tests
drwx------   1 856065   800588          4 09 Sep 12:33 l2
drwx------   1 856065   800588          4 14 Dec 2021  ls6
-rw-------   1 856065   800588          5 07 Feb 22:31 upload_test.txt
drwx------   1 856065   800588          9 12 May 2022  temp
drwx------   1 856065   800588          7 02 Dec 2020  si-base
drwx------   1 856065   800588          4 31 Jan 16:28 ch-sim-ls6
drwx------   1 856065   800588          4 11 Jan 2022  test
drwx------   1 856065   800588          2 30 Jan 15:54 ch-sim
drwx------   1 856065   800588          1 09 Sep 18:29 ios-output
drwx------   1 856065   800588          3 07 Feb 22:31 test-upload
-rw-------   1 856065   800588         12 07 Feb 22:15 hello.txt
-rw-------   1 856065   800588         36 07 Feb

In [12]:
print(tsc.execute_command("ls -lt $SCRATCH")["stdout"])

total 8
drwx------ 2 clos21 G-800588 1 Feb  2 12:00 ch-sim-tests
drwx------ 6 clos21 G-800588 4 Jan 31 16:28 ch-sim-ls6
drwx------ 4 clos21 G-800588 2 Jan 30 15:54 ch-sim
drwx------ 6 clos21 G-800588 4 Sep 30 20:52 test_ls6
drwx------ 3 clos21 G-800588 1 Sep  9 18:29 ios-output
drwx------ 6 clos21 G-800588 4 Sep  9 12:33 l2
drwx------ 6 clos21 G-800588 4 Sep  8 22:48 test-taccjm
drwx------ 6 clos21 G-800588 4 Aug 25 11:21 ls1
drwx------ 6 clos21 G-800588 4 Aug 19 05:48 l1
drwx------ 6 clos21 G-800588 4 Aug 19 05:32 11
drwx------ 6 clos21 G-800588 4 Jun  3  2022 taccjm
drwx------ 2 clos21 G-800588 9 May 12  2022 temp
drwx------ 6 clos21 G-800588 4 Jan 11  2022 test
drwx------ 6 clos21 G-800588 4 Dec 14  2021 ls6
drwx------ 3 clos21 G-800588 7 Dec  2  2020 si-base



### `upload()`

To upload a local directory or file to TACC, use the upload method.

In [38]:
!rm -rf test; mkdir test; echo test > test/test.txt

In [39]:
tsc.upload("test/test.txt", "test.txt")

In [40]:
tsc.list_files("test.txt")

[{'filename': 'test.txt',
  'st_atime': 1675808371,
  'st_gid': 800588,
  'st_mode': 33152,
  'st_mtime': 1675808371,
  'st_size': 5,
  'st_uid': 856065,
  'ls_str': b'-rw-------   1 856065   800588          5 07 Feb 16:19 ?'}]

In [43]:
tsc.upload("test", "test")

In [44]:
tsc.list_files("test")

[{'filename': 'test.txt',
  'st_atime': 1675808477,
  'st_gid': 800588,
  'st_mode': 33152,
  'st_mtime': 1675808368,
  'st_size': 5,
  'st_uid': 856065,
  'ls_str': b'-rw-------   1 856065   800588          5 07 Feb 16:19 test.txt'}]

### `download()`

In [45]:
tsc.download("test.txt", "foo.txt")

In [47]:
!cat foo.txt

test


In [51]:
tsc.download("test", "foo")

In [52]:
!ls foo/test

test.txt


### `write()`

In [54]:
tsc.write("hello world!\n", "test2.txt")

In [55]:
tsc.list_files("test2.txt")

[{'filename': 'test2.txt',
  'st_atime': 1675808673,
  'st_gid': 800588,
  'st_mode': 33152,
  'st_mtime': 1675808673,
  'st_size': 13,
  'st_uid': 856065,
  'ls_str': b'-rw-------   1 856065   800588         13 07 Feb 16:24 ?'}]

### `read()`

In [56]:
msg = tsc.read("test2.txt")
print(msg)

hello world!



# TACC SSH API - tacc_ssh_api.py

The problem with managing individual TACCSSHClient classes is that they may time out after a period of inactivity, requiring you to login again.
Instead of using the TACCSSHClient class directly, you can use the `tacc_ssh_api` to spawn a server that manages TACCSSHClient classes for you.
These connections are kept alive (for as long as the server isn't restarted), and they can be interacted via methods in `tacc_ssh_api`.
The main methods provided to interact with the SSH server are the following:

- `list()` - List active ssh session
- `get(connection_id: int)` - Get a session an active session
- `init(...)` - Initialize an ssh session (by initializing a new TACCSSHClient object)
- `exec(connection_id:int, ...)` - Run a command using `execute_command()` on a given session.
- `process(connection_id:int, cmnd_id:int ...)` - Process a command run on a given session.
- `list_files(connection_id:int, path:str ...)` - List files.
- `read(connection_id:int, path:str ...)` - Read a file directly via sftp.
- `write(connection_id:int, data, path:str)` - Write a file directly via sftp.
- `upload(connection_id:int, local_path:str, remote_path:str)` - Upload files/folders
- `download(connection_id:int, data, remote_path:str, local_path:Str)` - Download files/folders

## Initializing a sessions

In [1]:
from taccjm import tacc_ssh_api as tsa

In [2]:
tsa.find_server(start=True, kill=True)

{'server': <Popen: returncode: None args: ['python', '/home/jovyan/work/repos/taccjm/sr...>,
 'hb': <Popen: returncode: None args: ['python', '/home/jovyan/work/repos/taccjm/sr...>}

In [3]:
tsa.init("test", "ls6", "clos21")

mfa:  683832


{'id': 'test',
 'sys': 'ls6.tacc.utexas.edu',
 'user': 'clos21',
 'start': '2023-02-09T13:22:41.691522',
 'last_ts': '2023-02-09T13:22:41.691538',
 'log_level': '20',
 'log_file': '/home/jovyan/.taccjm/test_ls6_clos21_log.json',
 'home_dir': '/home1/06307/clos21',
 'work_dir': '/work/06307/clos21/ls6',
 'scratch_dir': '/scratch/06307/clos21'}

In [4]:
tsa.get("test")

{'id': 'test',
 'sys': 'ls6.tacc.utexas.edu',
 'user': 'clos21',
 'start': '2023-02-07T22:35:07.435852',
 'last_ts': '2023-02-07T22:35:07.435853',
 'loglevel': '20',
 'logfile': '/home/jovyan/.taccjm/test_ls6_clos21_log.json'}

## Command Operations

### `exec()` - General Command Execution

...

In [5]:
ls_cmnd = tsa.exec("test", "ls $SCRATCH", wait=False)

### `process()` - Polling commands

Pass a command ID to poll a specific command. 
Set `wait=False`, unless you want to wait until the command finishes.
You can set `nbytes=int` to get the latest nbytes from the stdout buff .

In [13]:
ls_cmnd = tsa.exec("test", "sleep 10; ls $SCRATCH", wait=False)

In [14]:
tsa.process("test", ls_cmnd["id"], wait=False)

{'id': 10,
 'cmd': 'sleep 10; ls $SCRATCH',
 'ts': '2023-02-07T18:22:49.523330',
 'status': 'RUNNING',
 'stdout': '',
 'stderr': '',
 'history': [{'ts': '2023-02-07T18:22:48.983625', 'status': 'STARTED'}],
 'rt': None}

### Processing all Active commands


In [5]:
cmnd_1 = tsa.exec("test", "sleep 10; ls $SCRATCH", wait=False)
cmnd_2 = tsa.exec("test", "sleep 20; ls $SCRATCH", wait=False)
cmnd_3 = tsa.exec("test", "sleep 30; ls $SCRATCH", wait=False)

In [14]:
tsa.process("test")

[{'id': 3,
  'cmd': 'sleep 20; ls $SCRATCH',
  'ts': '2023-02-07T18:28:00.670353',
  'status': 'RUNNING',
  'stdout': '',
  'stderr': '',
  'history': [{'ts': '2023-02-07T18:27:43.070502', 'status': 'STARTED'}],
  'rt': None},
 {'id': 4,
  'cmd': 'sleep 30; ls $SCRATCH',
  'ts': '2023-02-07T18:28:00.670360',
  'status': 'RUNNING',
  'stdout': '',
  'stderr': '',
  'history': [{'ts': '2023-02-07T18:27:43.146205', 'status': 'STARTED'}],
  'rt': None}]

## File Operations

### `list_files()`

The `list_files()` method provides basic file stat info, much like an `ls` command would in a shell session.

In [6]:
scratch_dir = tsa.list_files("test", "")
scratch_dir 

[{'filename': '/scratch/06307/clos21/',
  'st_atime': 1634240870,
  'st_gid': 800588,
  'st_mode': 16832,
  'st_mtime': 1675831233,
  'st_size': 19,
  'st_uid': 856065,
  'ls_str': 'drwx------   1 856065   800588         19 07 Feb 22:40 ?'}]

In [7]:
files = tsa.list_files("test", "", recurse=True)
len(files)

19

In [8]:
print("\n".join([x["ls_str"] for x in files if not x["filename"].startswith(".")]))

drwx------   1 856065   800588          4 19 Aug 05:32 11
drwx------   1 856065   800588          4 03 Jun 2022  taccjm
drwx------   1 856065   800588          4 08 Sep 22:48 test-taccjm
drwx------   1 856065   800588          1 02 Feb 12:00 ch-sim-tests
drwx------   1 856065   800588          4 09 Sep 12:33 l2
drwx------   1 856065   800588          4 14 Dec 2021  ls6
-rw-------   1 856065   800588          5 07 Feb 22:31 upload_test.txt
drwx------   1 856065   800588          9 12 May 2022  temp
drwx------   1 856065   800588          7 02 Dec 2020  si-base
drwx------   1 856065   800588          4 31 Jan 16:28 ch-sim-ls6
drwx------   1 856065   800588          4 11 Jan 2022  test
drwx------   1 856065   800588          2 30 Jan 15:54 ch-sim
drwx------   1 856065   800588          1 09 Sep 18:29 ios-output
drwx------   1 856065   800588          3 07 Feb 22:31 test-upload
-rw-------   1 856065   800588         12 07 Feb 22:15 hello.txt
-rw-------   1 856065   800588         36 07 Feb

In [9]:
print(tsa.exec("test", "ls -lt $SCRATCH")["stdout"])

total 10
drwx------ 2 clos21 G-800588  3 Feb  7 22:31 test-upload
-rw------- 1 clos21 G-800588  5 Feb  7 22:31 upload_test.txt
-rw------- 1 clos21 G-800588 36 Feb  7 22:17 hello.json
-rw------- 1 clos21 G-800588 12 Feb  7 22:15 hello.txt
drwx------ 2 clos21 G-800588  1 Feb  2 12:00 ch-sim-tests
drwx------ 6 clos21 G-800588  4 Jan 31 16:28 ch-sim-ls6
drwx------ 4 clos21 G-800588  2 Jan 30 15:54 ch-sim
drwx------ 6 clos21 G-800588  4 Sep 30 20:52 test_ls6
drwx------ 3 clos21 G-800588  1 Sep  9 18:29 ios-output
drwx------ 6 clos21 G-800588  4 Sep  9 12:33 l2
drwx------ 6 clos21 G-800588  4 Sep  8 22:48 test-taccjm
drwx------ 6 clos21 G-800588  4 Aug 25 11:21 ls1
drwx------ 6 clos21 G-800588  4 Aug 19 05:48 l1
drwx------ 6 clos21 G-800588  4 Aug 19 05:32 11
drwx------ 6 clos21 G-800588  4 Jun  3  2022 taccjm
drwx------ 2 clos21 G-800588  9 May 12  2022 temp
drwx------ 6 clos21 G-800588  4 Jan 11  2022 test
drwx------ 6 clos21 G-800588  4 Dec 14  2021 ls6
drwx------ 3 clos21 G-800588  7 Dec

### `write()` and `read()`

Use to write directly to a file on TACC systems.

#### Text Data

In [4]:
tsa.write("test", "hello world!", "hello.txt")

In [5]:
tsa.read("test", "hello.txt")

{'path': 'hello.txt', 'data_type': 'text', 'data': 'hello world!'}

#### json Data

In [7]:
tsa.write("test", {"msg": "hello world!", "id": 12312}, "hello.json")

In [8]:
tsa.read("test", "hello.json")

{'path': 'hello.json',
 'data_type': 'json',
 'data': {'msg': 'hello world!', 'id': 12312}}

### `upload()` and `download()`

To upload a local directory or file to TACC, use the upload method.

In [6]:
!rm -rf test-upload; mkdir test-upload; echo test > test-upload/test.txt

#### Files

In [5]:
tsa.upload("test", "test-upload/test.txt", "upload_test.txt")

In [6]:
tsa.list_files("test", "upload_test.txt")

[{'filename': '/scratch/06307/clos21/upload_test.txt',
  'st_atime': 1675830682,
  'st_gid': 800588,
  'st_mode': 33152,
  'st_mtime': 1675830682,
  'st_size': 5,
  'st_uid': 856065,
  'ls_str': '-rw-------   1 856065   800588          5 07 Feb 22:31 ?'}]

In [7]:
tsa.download("test", "upload_test.txt", "test-upload/download_test.txt")

'test-upload/download_test.txt'

In [8]:
!cat test-upload/download_test.txt

test


#### Directories

In [5]:
!echo 'another test' > test-upload/test2.txt

In [4]:
tsa.upload("test", "test-upload", "test-upload")

In [5]:
tsa.list_files("test", "test-upload")

[{'filename': 'download_test.txt',
  'st_atime': 1675831198,
  'st_gid': 800588,
  'st_mode': 33152,
  'st_mtime': 1675830687,
  'st_size': 5,
  'st_uid': 856065,
  'ls_str': '-rw-------   1 856065   800588          5 07 Feb 22:31 download_test.txt'},
 {'filename': 'test2.txt',
  'st_atime': 1675831199,
  'st_gid': 800588,
  'st_mode': 33152,
  'st_mtime': 1675830915,
  'st_size': 13,
  'st_uid': 856065,
  'ls_str': '-rw-------   1 856065   800588         13 07 Feb 22:35 test2.txt'},
 {'filename': 'test.txt',
  'st_atime': 1675831199,
  'st_gid': 800588,
  'st_mode': 33152,
  'st_mtime': 1675830340,
  'st_size': 5,
  'st_uid': 856065,
  'ls_str': '-rw-------   1 856065   800588          5 07 Feb 22:25 test.txt'}]

In [6]:
tsa.download("test", "test-upload", "test-download")

'test-download'

In [8]:
!ls -R test-download

test-download:
test-upload

test-download/test-upload:
download_test.txt  test2.txt  test.txt


# TACC Client


In [26]:
import importlib
import logging
import os
import sys
from pathlib import Path

from taccjm import TACCClient

In [27]:
importlib.reload(TACCClient)

<module 'taccjm.TACCClient' from '/home/jovyan/work/repos/taccjm/src/taccjm/TACCClient.py'>

## Remote Connections

If we aren't running on TACC, SSH session will be started.

In [29]:
os.environ["HOSTNAME"] = ""

In [30]:
client = TACCClient.TACCClient(system='stampede2')

In [31]:
client.scratch_dir, client.home_dir, client.work_dir,

('/scratch/06307/clos21',
 '/home1/06307/clos21',
 '/work2/06307/clos21/stampede2')

In [32]:
client.jobs_dir, client.scripts_dir, client.apps_dir, client.trash_dir

('/scratch/06307/clos21/taccjm-stampede2/jobs',
 '/work2/06307/clos21/stampede2/taccjm-stampede2/scripts',
 '/work2/06307/clos21/stampede2/taccjm-stampede2/apps',
 '/scratch/06307/clos21/taccjm-stampede2/trash')

In [33]:
scratch_dir = client.list_files()
scratch_dir

[{'filename': '/scratch/06307/clos21/',
  'st_atime': 1676065094,
  'st_gid': 821083,
  'st_mode': 16837,
  'st_mtime': 1675987758,
  'st_size': 4096,
  'st_uid': 856065,
  'ls_str': 'drwx---r-x   1 856065   821083       4096 09 Feb 18:09 ?'}]

In [34]:
files = client.list_files(recurse=True)
len(files)

29

### Mamba Environments

In [35]:
res = client.mamba_install('taccjm', 'pip', 'taccjm')

In [42]:
import pandas as pd
from io import StringIO

['# Name                    Version                   Build  Channel', '_libgcc_mutex             0.1                        main  ', 'ca-certificates           2020.1.1                      0  ', 'certifi                   2020.4.5.1               py38_0  ', 'cffi                      1.14.0           py38he30daa8_1  ', 'chardet                   3.0.4                 py38_1003  ', 'conda                     4.8.3                    py38_0  ', 'conda-package-handling    1.6.1            py38h7b6447c_0  ', 'cryptography              2.9.2            py38h1ba5d50_0  ', 'idna                      2.9                        py_1  ', 'ld_impl_linux-64          2.33.1               h53a641e_7  ', 'libedit                   3.1.20181209         hc058e9b_0  ', 'libffi                    3.3                  he6710b0_1  ', 'libgcc-ng                 9.1.0                hdf63c60_0  ', 'libstdcxx-ng              9.1.0                hdf63c60_0  ', 'ncurses                   6.2                 

In [63]:
output = client.exec('conda activate taccjm && conda list')['stdout']
print(output)

# packages in environment at /home1/06307/clos21/miniconda3/envs/taccjm:
#
# Name                    Version                   Build  Channel
_libgcc_mutex             0.1                        main  
_openmp_mutex             5.1                       1_gnu  
bcrypt                    4.0.1                    pypi_0    pypi
begins                    0.9                      pypi_0    pypi
bzip2                     1.0.8                h7b6447c_0  
ca-certificates           2023.01.10           h06a4308_0  
certifi                   2022.12.7                pypi_0    pypi
cffi                      1.15.1                   pypi_0    pypi
charset-normalizer        3.0.1                    pypi_0    pypi
click                     8.1.3                    pypi_0    pypi
cryptography              39.0.1                   pypi_0    pypi
falcon                    2.0.0                    pypi_0    pypi
hug                       2.6.1                    pypi_0    pypi
idna                    

In [65]:
df = pd.read_csv(StringIO(client.exec('conda activate taccjm && conda list')['stdout']), delim_whitespace=True, names=['name', 'version', 'build', 'channel'], header=None, skiprows=3)
df

name     version         build channel
0        _libgcc_mutex         0.1          main     NaN
1        _openmp_mutex         5.1         1_gnu     NaN
2               bcrypt       4.0.1        pypi_0    pypi
3               begins         0.9        pypi_0    pypi
4                bzip2       1.0.8    h7b6447c_0     NaN
5      ca-certificates  2023.01.10    h06a4308_0     NaN
6              certifi   2022.12.7        pypi_0    pypi
7                 cffi      1.15.1        pypi_0    pypi
8   charset-normalizer       3.0.1        pypi_0    pypi
9                click       8.1.3        pypi_0    pypi
10        cryptography      39.0.1        pypi_0    pypi
11              falcon       2.0.0        pypi_0    pypi
12                 hug       2.6.1        pypi_0    pypi
13                idna         3.4        pypi_0    pypi
14              jinja2       3.1.2        pypi_0    pypi
15    ld_impl_linux-64        2.38    h1181459_1     NaN
16              libffi       3.4.2    h6a678d5_6     NaN
17           libgcc-ng      11.2.0    h1234567_1     NaN
18             libgomp      11.2.0    h1234567_1     NaN
19        libstdcxx-ng      11.2.0    h1234567_1     NaN
20             libuuid      1.41.5    h5eee18b_0     NaN
21          markupsafe       2.1.2        pypi_0    pypi
22             ncurses         6.4    h6a678d5_0     NaN
23               numpy      1.24.2        pypi_0    pypi
24             openssl      1.1.1s    h7f8727e_0     NaN
25            paramiko       3.0.0        pypi_0    pypi
26                 pip      22.3.1        pypi_0    pypi
27         prettytable       3.6.0        pypi_0    pypi
28              psutil       5.9.4        pypi_0    pypi
29           pycparser        2.21        pypi_0    pypi
30              pynacl       1.5.0        pypi_0    pypi
31              python      3.10.9    h7a1cb2a_0     NaN
32  python-json-logger       2.0.4        pypi_0    pypi
33            readline         8.2    h5eee18b_0     NaN
34            requests      2.28.2        pypi_0    pypi
35          setuptools      65.6.3        pypi_0    pypi
36              sqlite      3.40.1    h5082296_0     NaN
37              taccjm       0.0.4        pypi_0    pypi
38                  tk      8.6.12    h1ccaba5_0     NaN
39              tzdata       2022g    h04d1e81_0     NaN
40             urllib3     1.26.14        pypi_0    pypi
41             wcwidth       0.2.6        pypi_0    pypi
42               wheel      0.37.1  pyhd3eb1b0_0     NaN
43                  xz      5.2.10    h5eee18b_1     NaN
44                zlib      1.2.13    h5eee18b_0     NaN

## Local TACC Sessions

Can mock being on TACC systems by setting the appropriate environment variables

In [21]:
import importlib
import logging
import os
import sys
from pathlib import Path

from taccjm import TACCClient

In [22]:
importlib.reload(TACCClient)

<module 'taccjm.TACCClient' from '/home/jovyan/work/repos/taccjm/src/taccjm/TACCClient.py'>

In [23]:
!mkdir data/test-scratch; mkdir data/test-home; mkdir data/test-work

mkdir: cannot create directory ‘data/test-scratch’: File exists
mkdir: cannot create directory ‘data/test-home’: File exists
mkdir: cannot create directory ‘data/test-work’: File exists


In [24]:
system = "ls6"
os.environ["HOSTNAME"] = f"login2.{system}.tacc.utexas.edu"
os.environ["SCRATCH"] = str(Path.cwd() / "data/test-scratch")
os.environ["HOME"] = str(Path.cwd() / "data/test-home")
os.environ["WORK"] = str(Path.cwd() / "data/test-work")
client = TACCClient.TACCClient(
    "ls6", log_config={'output': sys.stdout, 'fmt': 'json', 'level': logging.INFO}
)
os.environ["HOSTNAME"] = ""
os.environ["SCRATCH"] = ""
os.environ["HOME"] = ""
os.environ["WORK"] = ""

{"asctime": "2023-02-11 10:05:21,446", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Logger taccjm.TACCClient initialized", "config": {"output": "<ipykernel.iostream.OutStream object at 0x7f97b452e770>", "fmt": "json", "level": 20}}
{"asctime": "2023-02-11 10:05:21,446", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Logger taccjm.TACCClient initialized", "config": {"output": "<ipykernel.iostream.OutStream object at 0x7f97b452e770>", "fmt": "json", "level": 20}}
{"asctime": "2023-02-11 10:05:21,446", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Logger taccjm.TACCClient initialized", "config": {"output": "<ipykernel.iostream.OutStream object at 0x7f97b452e770>", "fmt": "json", "level": 20}}
{"asctime": "2023-02-11 10:05:21,446", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Logger taccjm.TACCClient initialized", "config": {"output": "<ipykernel.iostream.OutStream object at 0x7f97b452e770>", "fmt": "json", "level": 20}}
{"as

In [12]:
client.apps_dir, client.jobs_dir, client.scripts_dir, client.jobs_dir, client.trash_dir

('/home/jovyan/work/repos/taccjm/notebooks/data/test-work/taccjm-ls6/apps',
 '/home/jovyan/work/repos/taccjm/notebooks/data/test-scratch/taccjm-ls6/jobs',
 '/home/jovyan/work/repos/taccjm/notebooks/data/test-work/taccjm-ls6/scripts',
 '/home/jovyan/work/repos/taccjm/notebooks/data/test-scratch/taccjm-ls6/jobs',
 '/home/jovyan/work/repos/taccjm/notebooks/data/test-scratch/taccjm-ls6/trash')

### Command Operations

In [26]:
client.exec("ls")

2023-02-08 16:16:07,429 | taccjm.TACCSimulation | INFO | Running command locally
2023-02-08 16:16:07,429 | taccjm.TACCSimulation | INFO | Running command locally
2023-02-08 16:16:07,429 | taccjm.TACCSimulation | INFO | Running command locally
2023-02-08 16:16:07,429 | taccjm.TACCSimulation | INFO | Running command locally
2023-02-08 16:16:07,429 | taccjm.TACCSimulation | INFO | Running command locally
2023-02-08 16:16:07,435 | taccjm.TACCSimulation | INFO | Getting command 0 from local command list
2023-02-08 16:16:07,435 | taccjm.TACCSimulation | INFO | Getting command 0 from local command list
2023-02-08 16:16:07,435 | taccjm.TACCSimulation | INFO | Getting command 0 from local command list
2023-02-08 16:16:07,435 | taccjm.TACCSimulation | INFO | Getting command 0 from local command list
2023-02-08 16:16:07,435 | taccjm.TACCSimulation | INFO | Getting command 0 from local command list
2023-02-08 16:16:07,439 | taccjm.TACCSimulation | INFO | Waiting for command 0 to finish....
2023-02

{'id': 0,
 'cmd': 'ls',
 'ts': datetime.datetime(2023, 2, 8, 16, 16, 7, 442880),
 'status': 'STARTED',
 'stdout': b'data\ntest.json\nTutorial.ipynb\n',
 'stderr': b'',
 'history': [{'status': 'STARTED',
   'ts': datetime.datetime(2023, 2, 8, 16, 16, 7, 435737)}],
 'process': <Popen: returncode: 0 args: 'ls'>,
 'rt': 0,
 'rc': 0}

In [27]:
client.exec("echo START; sleep 10; echo END", wait=False)

2023-02-08 16:17:46,454 | taccjm.TACCSimulation | INFO | Running command locally
2023-02-08 16:17:46,454 | taccjm.TACCSimulation | INFO | Running command locally
2023-02-08 16:17:46,454 | taccjm.TACCSimulation | INFO | Running command locally
2023-02-08 16:17:46,454 | taccjm.TACCSimulation | INFO | Running command locally
2023-02-08 16:17:46,454 | taccjm.TACCSimulation | INFO | Running command locally


{'id': 1,
 'cmd': 'echo START; sleep 10; echo END',
 'ts': datetime.datetime(2023, 2, 8, 16, 17, 46, 461570),
 'status': 'STARTED',
 'stdout': '',
 'stderr': '',
 'history': [],
 'process': <Popen: returncode: None args: 'echo START; sleep 10; echo END'>,
 'rt': None}

In [30]:
client.process(1)

2023-02-08 16:18:09,027 | taccjm.TACCSimulation | INFO | Getting command 1 from local command list
2023-02-08 16:18:09,027 | taccjm.TACCSimulation | INFO | Getting command 1 from local command list
2023-02-08 16:18:09,027 | taccjm.TACCSimulation | INFO | Getting command 1 from local command list
2023-02-08 16:18:09,027 | taccjm.TACCSimulation | INFO | Getting command 1 from local command list
2023-02-08 16:18:09,027 | taccjm.TACCSimulation | INFO | Getting command 1 from local command list
2023-02-08 16:18:09,032 | taccjm.TACCSimulation | INFO | Waiting for command 1 to finish....
2023-02-08 16:18:09,032 | taccjm.TACCSimulation | INFO | Waiting for command 1 to finish....
2023-02-08 16:18:09,032 | taccjm.TACCSimulation | INFO | Waiting for command 1 to finish....
2023-02-08 16:18:09,032 | taccjm.TACCSimulation | INFO | Waiting for command 1 to finish....
2023-02-08 16:18:09,032 | taccjm.TACCSimulation | INFO | Waiting for command 1 to finish....


{'id': 1,
 'cmd': 'echo START; sleep 10; echo END',
 'ts': datetime.datetime(2023, 2, 8, 16, 18, 9, 36562),
 'status': 'STARTED',
 'stdout': b'START\nEND\n',
 'stderr': b'',
 'history': [{'status': 'STARTED',
   'ts': datetime.datetime(2023, 2, 8, 16, 17, 46, 461570)}],
 'process': <Popen: returncode: 0 args: 'echo START; sleep 10; echo END'>,
 'rt': 22,
 'rc': 0}

### File Operations

Note: Upload and download don't really mean much when executing on TACC itself. 

In [12]:
current_dir = client.list_files("")
current_dir

[[{'filename': '/home/jovyan/work/repos/taccjm/notebooks',
   'st_atime': 1675972885.788579,
   'st_gid': 100,
   'st_mode': 17901,
   'st_mtime': 1675972885.753579,
   'st_size': 4096,
   'st_uid': 1000}]]

In [15]:
files = client.list_files("", recurse=True)
len(files)

10

In [12]:
client.write({"test": 1231}, "test.json")

In [13]:
client.list_files("test.json")

[{'filename': '/home/jovyan/work/repos/taccjm/notebooks/test.json',
  'st_atime': 1675972951.953107,
  'st_gid': 100,
  'st_mode': 33188,
  'st_mtime': 1675977509.267336,
  'st_size': 14,
  'st_uid': 1000}]

In [14]:
client.read("test.json")

'{"test": 1231}'

In [23]:
client.list_files('../../..')

[[{'filename': '/home/jovyan/work',
   'st_atime': 1675970669.301117,
   'st_gid': 100,
   'st_mode': 19965,
   'st_mtime': 1675966498.928591,
   'st_size': 4096,
   'st_uid': 1000}]]

### Trash operations

In [4]:
client.trash_dir

'/home/jovyan/work/repos/taccjm/notebooks/data/test-scratch/taccjm-ls6/trash'

In [4]:
client.ls_trash()

[]

In [5]:
client.rm('test.json')

{"asctime": "2023-02-09 15:34:52,791", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Running command locally", "cmnd": "rsync -a /home/jovyan/work/repos/taccjm/notebooks/test.json /home/jovyan/work/repos/taccjm/notebooks/data/test-scratch/taccjm-ls6/trash/___home___jovyan___work___repos___taccjm___notebooks___test.json && rm -rf /home/jovyan/work/repos/taccjm/notebooks/test.json", "wait": false, "error": true}


{'id': 4,
 'cmd': 'rsync -a /home/jovyan/work/repos/taccjm/notebooks/test.json /home/jovyan/work/repos/taccjm/notebooks/data/test-scratch/taccjm-ls6/trash/___home___jovyan___work___repos___taccjm___notebooks___test.json && rm -rf /home/jovyan/work/repos/taccjm/notebooks/test.json',
 'ts': datetime.datetime(2023, 2, 9, 15, 34, 52, 792784),
 'status': 'STARTED',
 'stdout': '',
 'stderr': '',
 'history': [],
 'process': <Popen: returncode: None args: 'rsync -a /home/jovyan/work/repos/taccjm/note...>,
 'rt': None}

In [6]:
client.process(4)

{"asctime": "2023-02-09 15:34:54,600", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Getting command 4 from local command list"}
{"asctime": "2023-02-09 15:34:54,603", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Waiting for command 4 to finish...."}
{"asctime": "2023-02-09 15:34:54,604", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Command 4 done. Reading output."}
{"asctime": "2023-02-09 15:34:54,605", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Command 4 completed!"}
{"asctime": "2023-02-09 15:34:54,606", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Processed command 4", "cmnd_config": {"id": 4, "cmd": "rsync -a /home/jovyan/work/repos/taccjm/notebooks/test.json /home/jovyan/work/repos/taccjm/notebooks/data/test-scratch/taccjm-ls6/trash/___home___jovyan___work___repos___taccjm___notebooks___test.json && rm -rf /home/jovyan/work/repos/taccjm/notebooks/test.json", "ts": "2023-02-09T15:34:54.605277", "

{'id': 4,
 'cmd': 'rsync -a /home/jovyan/work/repos/taccjm/notebooks/test.json /home/jovyan/work/repos/taccjm/notebooks/data/test-scratch/taccjm-ls6/trash/___home___jovyan___work___repos___taccjm___notebooks___test.json && rm -rf /home/jovyan/work/repos/taccjm/notebooks/test.json',
 'ts': datetime.datetime(2023, 2, 9, 15, 34, 54, 605277),
 'status': 'COMPLETE',
 'stdout': b'',
 'stderr': b'',
 'history': [{'status': 'STARTED',
   'ts': datetime.datetime(2023, 2, 9, 15, 34, 52, 792784)}],
 'process': <Popen: returncode: 0 args: 'rsync -a /home/jovyan/work/repos/taccjm/noteboo...>,
 'rt': 1,
 'rc': 0}

In [13]:
client.ls_trash()

[{'filename': '/home/jovyan/work/repos/taccjm/notebooks/test.json',
  'st_atime': 1675978492.799484,
  'st_gid': 100,
  'st_mode': 33188,
  'st_mtime': 1675977509.267336,
  'st_size': 14,
  'st_uid': 1000}]

In [14]:
client.list_files("test.json")

ValueError: Error getting stats for /home/jovyan/work/repos/taccjm/notebooks/test.json: [Errno 2] No such file or directory: '/home/jovyan/work/repos/taccjm/notebooks/test.json'

In [15]:
client.rm('test.json', restore=True)

{"asctime": "2023-02-09 15:36:16,137", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Running command locally", "cmnd": "rsync -a /home/jovyan/work/repos/taccjm/notebooks/data/test-scratch/taccjm-ls6/trash/___home___jovyan___work___repos___taccjm___notebooks___test.json /home/jovyan/work/repos/taccjm/notebooks/test.json && rm -rf /home/jovyan/work/repos/taccjm/notebooks/data/test-scratch/taccjm-ls6/trash/___home___jovyan___work___repos___taccjm___notebooks___test.json", "wait": false, "error": true}
{"asctime": "2023-02-09 15:36:16,137", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Running command locally", "cmnd": "rsync -a /home/jovyan/work/repos/taccjm/notebooks/data/test-scratch/taccjm-ls6/trash/___home___jovyan___work___repos___taccjm___notebooks___test.json /home/jovyan/work/repos/taccjm/notebooks/test.json && rm -rf /home/jovyan/work/repos/taccjm/notebooks/data/test-scratch/taccjm-ls6/trash/___home___jovyan___work___repos___taccjm___notebooks___te

{'id': 4,
 'cmd': 'rsync -a /home/jovyan/work/repos/taccjm/notebooks/data/test-scratch/taccjm-ls6/trash/___home___jovyan___work___repos___taccjm___notebooks___test.json /home/jovyan/work/repos/taccjm/notebooks/test.json && rm -rf /home/jovyan/work/repos/taccjm/notebooks/data/test-scratch/taccjm-ls6/trash/___home___jovyan___work___repos___taccjm___notebooks___test.json',
 'ts': datetime.datetime(2023, 2, 9, 15, 36, 16, 139573),
 'status': 'STARTED',
 'stdout': '',
 'stderr': '',
 'history': [],
 'process': <Popen: returncode: None args: 'rsync -a /home/jovyan/work/repos/taccjm/note...>,
 'rt': None}

In [16]:
client.list_files("test.json")

[{'filename': '/home/jovyan/work/repos/taccjm/notebooks/test.json',
  'st_atime': 1675978576.145376,
  'st_gid': 100,
  'st_mode': 33188,
  'st_mtime': 1675977509.267336,
  'st_size': 14,
  'st_uid': 1000}]

In [17]:
client.read('test.json')

'{"test": 1231}'

## Mamba/Conda Environments

Commands to handle environment set-up on tacc

In [25]:
envs = client.get_python_env('test')
envs

{"asctime": "2023-02-11 10:05:26,075", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Running command locally", "cmnd": "mamba --help", "wait": false, "error": true}
{"asctime": "2023-02-11 10:05:26,075", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Running command locally", "cmnd": "mamba --help", "wait": false, "error": true}
{"asctime": "2023-02-11 10:05:26,075", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Running command locally", "cmnd": "mamba --help", "wait": false, "error": true}
{"asctime": "2023-02-11 10:05:26,075", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Running command locally", "cmnd": "mamba --help", "wait": false, "error": true}
{"asctime": "2023-02-11 10:05:26,081", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Running command locally", "cmnd": "conda --help", "wait": false, "error": true}
{"asctime": "2023-02-11 10:05:26,081", "name": "taccjm.TACCClient", "levelname": "INFO", "messa

name    version               build      channel  \
0      _libgcc_mutex        0.1         conda_forge  conda-forge   
1      _openmp_mutex        4.5               2_gnu  conda-forge   
2              bzip2      1.0.8          h7f98852_4  conda-forge   
3    ca-certificates  2022.12.7          ha878542_0  conda-forge   
4   ld_impl_linux-64       2.40          h41732ed_0  conda-forge   
5             libffi      3.4.2          h7f98852_5  conda-forge   
6          libgcc-ng     12.2.0         h65d4601_19  conda-forge   
7            libgomp     12.2.0         h65d4601_19  conda-forge   
8             libnsl      2.0.0          h7f98852_0  conda-forge   
9          libsqlite     3.40.0          h753d276_0  conda-forge   
10           libuuid     2.32.1       h7f98852_1000  conda-forge   
11           libzlib     1.2.13          h166bdaf_4  conda-forge   
12           ncurses        6.3          h27087fc_1  conda-forge   
13           openssl      3.0.8          h0b41bf4_0  conda-forge   
14               pip       23.0        pyhd8ed1ab_0  conda-forge   
15            python     3.11.0  he550d4f_1_cpython  conda-forge   
16          readline      8.1.2          h0f457ee_0  conda-forge   
17        setuptools     67.1.0        pyhd8ed1ab_0  conda-forge   
18                tk     8.6.12          h27826a3_0  conda-forge   
19            tzdata      2022g          h191b570_0  conda-forge   
20             wheel     0.38.4        pyhd8ed1ab_0  conda-forge   
21                xz      5.2.6          h166bdaf_0  conda-forge   

                    path  
0   /opt/conda/envs/test  
1   /opt/conda/envs/test  
2   /opt/conda/envs/test  
3   /opt/conda/envs/test  
4   /opt/conda/envs/test  
5   /opt/conda/envs/test  
6   /opt/conda/envs/test  
7   /opt/conda/envs/test  
8   /opt/conda/envs/test  
9   /opt/conda/envs/test  
10  /opt/conda/envs/test  
11  /opt/conda/envs/test  
12  /opt/conda/envs/test  
13  /opt/conda/envs/test  
14  /opt/conda/envs/test  
15  /opt/conda/envs/test  
16  /opt/conda/envs/test  
17  /opt/conda/envs/test  
18  /opt/conda/envs/test  
19  /opt/conda/envs/test  
20  /opt/conda/envs/test  
21  /opt/conda/envs/test

In [11]:
fp = StringIO(output)

{"asctime": "2023-02-10 20:20:15,285", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Running command locally", "cmnd": "which pip", "wait": true, "error": true}
{"asctime": "2023-02-10 20:20:15,285", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Running command locally", "cmnd": "which pip", "wait": true, "error": true}
{"asctime": "2023-02-10 20:20:15,289", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Getting command 6 from local command list"}
{"asctime": "2023-02-10 20:20:15,289", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Getting command 6 from local command list"}
{"asctime": "2023-02-10 20:20:15,290", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Waiting for command 6 to finish...."}
{"asctime": "2023-02-10 20:20:15,290", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Waiting for command 6 to finish...."}
{"asctime": "2023-02-10 20:20:15,293", "name": "taccjm.TACCClient", "levelname

b'/opt/conda/bin/pip\n'

In [18]:
envs['path'][envs['name'] == 'test'].iloc[0]

'/opt/conda/envs/test'

In [45]:
import pandas as pd
print(envs.decode('utf-8'))

# conda environments:
#
base                     /opt/conda
test                     /opt/conda/envs/test




In [5]:
res = client

{"asctime": "2023-02-10 17:48:21,717", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Running command locally", "cmnd": "mamba --help", "wait": false, "error": true}
{"asctime": "2023-02-10 17:48:21,719", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Running command locally", "cmnd": "conda --help", "wait": false, "error": true}
{"asctime": "2023-02-10 17:48:21,721", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Getting command 2 from local command list"}
{"asctime": "2023-02-10 17:48:21,722", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Waiting for command 2 to finish...."}
{"asctime": "2023-02-10 17:48:22,190", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Command 2 done. Reading output."}
{"asctime": "2023-02-10 17:48:22,191", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Command 2 completed!"}
{"asctime": "2023-02-10 17:48:22,192", "name": "taccjm.TACCClient", "levelname": "INFO", "messa

TACCCommandError: Non-zero return code.
b'jovyan'@ls6$ mamba activate taccjm && mamba list
rc     : 255
stdout : b''
stderr : b"Run 'mamba init' to be able to run mamba activate/deactivate\nand start a new shell session. Or use conda to activate/deactivate.\n\n"

# TACC Base Simulation


## Local Session - Testing

In [1]:
import os
from pathlib import Path
from taccjm import TACCSimulation
import logging

In [2]:
# Move TACC System

def mock_tacc():
    system = "ls6"
    os.environ["HOSTNAME"] = f"login2.{system}.tacc.utexas.edu"
    os.environ["SCRATCH"] = str(Path.cwd() / "data/test-scratch")
    os.environ["HOME"] = str(Path.cwd() / "data/test-home")
    os.environ["WORK"] = str(Path.cwd() / "data/test-work")

def unmock_tacc():
    os.environ["HOSTNAME"] = ""
    os.environ["SCRATCH"] = ""
    os.environ["HOME"] = ""
    os.environ["WORK"] = ""

In [3]:
import importlib
importlib.reload(TACCSimulation)

<module 'taccjm.TACCSimulation' from '/home/jovyan/work/repos/taccjm/src/taccjm/TACCSimulation.py'>

In [4]:
mock_tacc()
sim = TACCSimulation.TACCSimulation(system='ls6',
                                    name='test',
                                    log_config={'output': 'local-test.log', 'level': logging.INFO})
unmock_tacc()

In [5]:
job_config = sim.setup(args={'input_file': 'data/test.txt', 'param': 5}, slurm_config={'processors_per_node': 1, 'allocation': 'ADCIRC'}, stage=True, run=False)
job_config

{'name': 'test',
 'job_id': 'test_20230211_204452qdkgxt0c',
 'job_dir': '/home/jovyan/work/repos/taccjm/notebooks/data/test-scratch/taccjm-ls6/jobs/test_20230211_204452qdkgxt0c',
 'slurm': {'allocation': 'ADCIRC',
  'node_count': 1,
  'processors_per_node': 1,
  'max_run_time': 0.1,
  'queue': 'development',
  'dependencies': []},
 'args': {'input_file': '/home/jovyan/work/repos/taccjm/notebooks/data/test-scratch/taccjm-ls6/jobs/test_20230211_204452qdkgxt0c/test.txt',
  'param': 5},
 'submit_script': '/tmp/tmp2p_jr201/submit_script.sh',
 'sim_script': '/tmp/tmp2p_jr201/test.py'}

In [10]:
sim._read_log()

asctime                   name levelname  \
0    2023-02-11 20:42:48,978  taccjm.TACCSimulation      INFO   
1    2023-02-11 20:42:48,978  taccjm.TACCSimulation      INFO   
2    2023-02-11 20:42:48,979      taccjm.TACCClient      INFO   
3    2023-02-11 20:42:48,979      taccjm.TACCClient      INFO   
4    2023-02-11 20:42:48,987      taccjm.TACCClient      INFO   
..                       ...                    ...       ...   
123  2023-02-11 20:44:52,072      taccjm.TACCClient      INFO   
124  2023-02-11 20:44:52,072      taccjm.TACCClient      INFO   
125  2023-02-11 20:44:52,120      taccjm.TACCClient      INFO   
126  2023-02-11 20:44:52,120      taccjm.TACCClient      INFO   
127  2023-02-11 20:44:52,120      taccjm.TACCClient      INFO   

                                               message  \
0             Logger taccjm.TACCSimulation initialized   
1    Running from non-main script /home/jovyan/work...   
2                 Logger taccjm.TACCClient initialized   
3                              Running command locally   
4            Getting command 0 from local command list   
..                                                 ...   
123          Getting command 7 from local command list   
124                Waiting for command 7 to finish....   
125                    Command 7 done. Reading output.   
126                               Command 7 completed!   
127                                Processed command 7   

                                                config    cmnd  wait  error  \
0    {'output': 'local-test.log', 'fmt': 'json', 'l...     NaN   NaN    NaN   
1                                                  NaN     NaN   NaN    NaN   
2    {'output': 'local-test.log', 'fmt': 'json', 'l...     NaN   NaN    NaN   
3                                                  NaN  whoami   1.0    1.0   
4                                                  NaN     NaN   NaN    NaN   
..                                                 ...     ...   ...    ...   
123                                                NaN     NaN   NaN    NaN   
124                                                NaN     NaN   NaN    NaN   
125                                                NaN     NaN   NaN    NaN   
126                                                NaN     NaN   NaN    NaN   
127                                                NaN     NaN   NaN    NaN   

                                           cmnd_config inputs slurm_config  \
0                                                  NaN    NaN          NaN   
1                                                  NaN    NaN          NaN   
2                                                  NaN    NaN          NaN   
3                                                  NaN    NaN          NaN   
4                                                  NaN    NaN          NaN   
..                                                 ...    ...          ...   
123                                                NaN    NaN          NaN   
124                                                NaN    NaN          NaN   
125                                                NaN    NaN          NaN   
126                                                NaN    NaN          NaN   
127  {'id': '7', 'cmd': 'rsync -a /tmp/tmp2p_jr201/...    NaN          NaN   

     arg run_cmd job_config job_dir  src dest path stage_commands  
0    NaN     NaN        NaN     NaN  NaN  NaN  NaN            NaN  
1    NaN     NaN        NaN     NaN  NaN  NaN  NaN            NaN  
2    NaN     NaN        NaN     NaN  NaN  NaN  NaN            NaN  
3    NaN     NaN        NaN     NaN  NaN  NaN  NaN            NaN  
4    NaN     NaN        NaN     NaN  NaN  NaN  NaN            NaN  
..   ...     ...        ...     ...  ...  ...  ...            ...  
123  NaN     NaN        NaN     NaN  NaN  NaN  NaN            NaN  
124  NaN     NaN        NaN     NaN  NaN  NaN  NaN            NaN  
125  NaN     NaN        NaN     NaN  NaN  NaN

In [7]:
sim.client.get_python_env(env='taccjm')

{"asctime": "2023-02-11 10:45:39,733", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Running command locally", "cmnd": "mamba list --name taccjm", "wait": true, "error": true}
{"asctime": "2023-02-11 10:45:39,737", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Getting command 8 from local command list"}
{"asctime": "2023-02-11 10:45:39,737", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Waiting for command 8 to finish...."}
{"asctime": "2023-02-11 10:45:40,332", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Command 8 done. Reading output."}
{"asctime": "2023-02-11 10:45:40,333", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Command 8 completed!"}
{"asctime": "2023-02-11 10:45:40,334", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Processed command 8", "cmnd_config": {"id": 8, "cmd": "mamba list --name taccjm", "ts": "2023-02-11T10:45:40.333318", "status": "COMPLETE", "stdout": "b'# packages 

name    version               build      channel  \
0      _libgcc_mutex        0.1         conda_forge  conda-forge   
1      _openmp_mutex        4.5               2_gnu  conda-forge   
2              bzip2      1.0.8          h7f98852_4  conda-forge   
3    ca-certificates  2022.12.7          ha878542_0  conda-forge   
4   ld_impl_linux-64       2.40          h41732ed_0  conda-forge   
5             libffi      3.4.2          h7f98852_5  conda-forge   
6          libgcc-ng     12.2.0         h65d4601_19  conda-forge   
7            libgomp     12.2.0         h65d4601_19  conda-forge   
8             libnsl      2.0.0          h7f98852_0  conda-forge   
9          libsqlite     3.40.0          h753d276_0  conda-forge   
10           libuuid     2.32.1       h7f98852_1000  conda-forge   
11           libzlib     1.2.13          h166bdaf_4  conda-forge   
12           ncurses        6.3          h27087fc_1  conda-forge   
13           openssl      3.0.8          h0b41bf4_0  conda-forge   
14               pip       23.0        pyhd8ed1ab_0  conda-forge   
15            python     3.11.0  he550d4f_1_cpython  conda-forge   
16          readline      8.1.2          h0f457ee_0  conda-forge   
17        setuptools     67.1.0        pyhd8ed1ab_0  conda-forge   
18                tk     8.6.12          h27826a3_0  conda-forge   
19            tzdata      2022g          h191b570_0  conda-forge   
20             wheel     0.38.4        pyhd8ed1ab_0  conda-forge   
21                xz      5.2.6          h166bdaf_0  conda-forge   

                      path  
0   /opt/conda/envs/taccjm  
1   /opt/conda/envs/taccjm  
2   /opt/conda/envs/taccjm  
3   /opt/conda/envs/taccjm  
4   /opt/conda/envs/taccjm  
5   /opt/conda/envs/taccjm  
6   /opt/conda/envs/taccjm  
7   /opt/conda/envs/taccjm  
8   /opt/conda/envs/taccjm  
9   /opt/conda/envs/taccjm  
10  /opt/conda/envs/taccjm  
11  /opt/conda/envs/taccjm  
12  /opt/conda/envs/taccjm  
13  /opt/conda/envs/taccjm  
14  /opt/conda/envs/taccjm  
15  /opt/conda/envs/taccjm  
16  /opt/conda/envs/taccjm  
17  /opt/conda/envs/taccjm  
18  /opt/conda/envs/taccjm  
19  /opt/conda/envs/taccjm  
20  /opt/conda/envs/taccjm  
21  /opt/conda/envs/taccjm

## Remote Session

In [1]:
import os
from pathlib import Path
from taccjm import TACCSimulation
import logging

In [17]:
import importlib
importlib.reload(TACCSimulation)

<module 'taccjm.TACCSimulation' from '/home/jovyan/work/repos/taccjm/src/taccjm/TACCSimulation.py'>

In [18]:
sim = TACCSimulation.TACCSimulation(system='ls6',
                                    name='test',
                                    log_config={'level': logging.INFO})

{"asctime": "2023-02-11 21:42:56,238", "name": "taccjm.TACCSimulation", "levelname": "INFO", "message": "Logger taccjm.TACCSimulation initialized", "config": {"output": "<ipykernel.iostream.OutStream object at 0x7f883fc027a0>", "fmt": "json", "level": 20}}
{"asctime": "2023-02-11 21:42:56,238", "name": "taccjm.TACCSimulation", "levelname": "INFO", "message": "Logger taccjm.TACCSimulation initialized", "config": {"output": "<ipykernel.iostream.OutStream object at 0x7f883fc027a0>", "fmt": "json", "level": 20}}
{"asctime": "2023-02-11 21:42:56,238", "name": "taccjm.TACCSimulation", "levelname": "INFO", "message": "Logger taccjm.TACCSimulation initialized", "config": {"output": "<ipykernel.iostream.OutStream object at 0x7f883fc027a0>", "fmt": "json", "level": 20}}
{"asctime": "2023-02-11 21:42:56,243", "name": "taccjm.TACCSimulation", "levelname": "INFO", "message": "Running from non-main script /home/jovyan/work/repos/taccjm/src/taccjm/TACCSimulation.py"}
{"asctime": "2023-02-11 21:42:56,

In [19]:
job_config = sim.run(args={'input_file': 'data/test.txt', 'param': 5}, slurm_config={'processors_per_node': 1, 'allocation': 'ADCIRC'})
job_config

{"asctime": "2023-02-11 21:42:57,898", "name": "taccjm.TACCSimulation", "levelname": "INFO", "message": "Not in execution environment. Setting up job..."}
{"asctime": "2023-02-11 21:42:57,898", "name": "taccjm.TACCSimulation", "levelname": "INFO", "message": "Not in execution environment. Setting up job..."}
{"asctime": "2023-02-11 21:42:57,898", "name": "taccjm.TACCSimulation", "levelname": "INFO", "message": "Not in execution environment. Setting up job..."}
{"asctime": "2023-02-11 21:42:57,901", "name": "taccjm.TACCSimulation", "levelname": "INFO", "message": "Starting simulation set-up.", "inputs": {"args": {"input_file": "data/test.txt", "param": 5}, "slurm_config": {"processors_per_node": 1, "allocation": "ADCIRC"}, "stage": true, "run": true}}
{"asctime": "2023-02-11 21:42:57,901", "name": "taccjm.TACCSimulation", "levelname": "INFO", "message": "Starting simulation set-up.", "inputs": {"args": {"input_file": "data/test.txt", "param": 5}, "slurm_config": {"processors_per_node": 

{'name': 'test',
 'job_id': 'test_20230211_214419qtt47a2m',
 'job_dir': '/scratch/06307/clos21/taccjm-ls6/jobs/test_20230211_214419qtt47a2m',
 'slurm': {'allocation': 'ADCIRC',
  'node_count': 1,
  'processors_per_node': 1,
  'max_run_time': 0.1,
  'queue': 'development',
  'dependencies': []},
 'args': {'input_file': '/scratch/06307/clos21/taccjm-ls6/jobs/test_20230211_214419qtt47a2m/test.txt',
  'param': 5},
 'submit_script': '/scratch/06307/clos21/taccjm-ls6/jobs/test_20230211_214419qtt47a2m/submit_script.sh',
 'sim_script': '/scratch/06307/clos21/taccjm-ls6/jobs/test_20230211_214419qtt47a2m/test.py',
 'slurm_id': '694729'}

In [20]:
files = sim.client.list_files('', job_id=job_config['job_id'], recurse=True)
files

[{'filename': 'test_20230211_214419qtt47a2m.e694729',
  'st_atime': 1676173482,
  'st_gid': 800588,
  'st_mode': 33152,
  'st_mtime': 1676173482,
  'st_size': 727,
  'st_uid': 856065,
  'ls_str': '-rw-------   1 856065   800588        727 11 Feb 21:44 test_20230211_214419qtt47a2m.e694729'},
 {'filename': 'submit_script.sh',
  'st_atime': 1676173464,
  'st_gid': 800588,
  'st_mode': 33152,
  'st_mtime': 1676173460,
  'st_size': 963,
  'st_uid': 856065,
  'ls_str': '-rw-------   1 856065   800588        963 11 Feb 21:44 submit_script.sh'},
 {'filename': 'test.txt',
  'st_atime': 1676173464,
  'st_gid': 800588,
  'st_mode': 33152,
  'st_mtime': 1676052351,
  'st_size': 24,
  'st_uid': 856065,
  'ls_str': '-rw-------   1 856065   800588         24 10 Feb 12:05 test.txt'},
 {'filename': 'test_20230211_214419qtt47a2m.o694729',
  'st_atime': 1676173470,
  'st_gid': 800588,
  'st_mode': 33152,
  'st_mtime': 1676173470,
  'st_size': 0,
  'st_uid': 856065,
  'ls_str': '-rw-------   1 856065   80

In [21]:
sim.client.download('', 'data', job_id=job_config['job_id'])

'data'

In [9]:
job_config = sim.client.submit_job(job_config['job_id'])
job_config 

TACCCommandError: Non-zero return code.
clos21@ls6$ cd /scratch/06307/clos21/taccjm-ls6/jobs/test_20230211_205944unt993hu && sbatch /tmp/tmp99th7z3o/submit_script.sh
rc     : 1
stdout : 
-----------------------------------------------------------------
          Welcome to the Lonestar6 Supercomputer          
-----------------------------------------------------------------


stderr : sbatch: error: Unable to open file /tmp/tmp99th7z3o/submit_script.sh


In [8]:
sim._read_log()

asctime                   name levelname  \
0    2023-02-11 20:42:48,978  taccjm.TACCSimulation      INFO   
1    2023-02-11 20:42:48,978  taccjm.TACCSimulation      INFO   
2    2023-02-11 20:42:48,979      taccjm.TACCClient      INFO   
3    2023-02-11 20:42:48,979      taccjm.TACCClient      INFO   
4    2023-02-11 20:42:48,987      taccjm.TACCClient      INFO   
..                       ...                    ...       ...   
350  2023-02-11 20:59:48,371      taccjm.TACCClient      INFO   
351  2023-02-11 20:59:48,417      taccjm.TACCClient      INFO   
352  2023-02-11 20:59:48,417      taccjm.TACCClient      INFO   
353  2023-02-11 20:59:48,417      taccjm.TACCClient      INFO   
354  2023-02-11 20:59:48,417  taccjm.TACCSimulation      INFO   

                                               message  \
0             Logger taccjm.TACCSimulation initialized   
1    Running from non-main script /home/jovyan/work...   
2                 Logger taccjm.TACCClient initialized   
3                              Running command locally   
4            Getting command 0 from local command list   
..                                                 ...   
350                Waiting for command 0 to finish....   
351                    Command 0 done. Reading output.   
352                               Command 0 completed!   
353                                Processed command 0   
354                            Uploading job directory   

                                                config    cmnd  wait  error  \
0    {'output': 'local-test.log', 'fmt': 'json', 'l...     NaN   NaN    NaN   
1                                                  NaN     NaN   NaN    NaN   
2    {'output': 'local-test.log', 'fmt': 'json', 'l...     NaN   NaN    NaN   
3                                                  NaN  whoami   1.0    1.0   
4                                                  NaN     NaN   NaN    NaN   
..                                                 ...     ...   ...    ...   
350                                                NaN     NaN   NaN    NaN   
351                                                NaN     NaN   NaN    NaN   
352                                                NaN     NaN   NaN    NaN   
353                                                NaN     NaN   NaN    NaN   
354                                                NaN     NaN   NaN    NaN   

                                           cmnd_config inputs  ...   id  sys  \
0                                                  NaN    NaN  ...  NaN  NaN   
1                                                  NaN    NaN  ...  NaN  NaN   
2                                                  NaN    NaN  ...  NaN  NaN   
3                                                  NaN    NaN  ...  NaN  NaN   
4                                                  NaN    NaN  ...  NaN  NaN   
..                                                 ...    ...  ...  ...  ...   
350                                                NaN    NaN  ...  NaN  NaN   
351                                                NaN    NaN  ...  NaN  NaN   
352                                                NaN    NaN  ...  NaN  NaN   
353  {'id': '0', 'cmd': 'rsync -a data/test.txt /tm...    NaN  ...  NaN  NaN   
354                                                NaN    NaN  ...  NaN  NaN   

    user start last_ts log_level log_file home_dir work_dir scratch_dir  
0    NaN   NaN     NaN       NaN      NaN      NaN      NaN         NaN  
1    NaN   NaN     NaN       NaN      NaN      NaN      NaN         NaN  
2    NaN   NaN     NaN       NaN      NaN      NaN      NaN         NaN  
3    NaN   NaN     NaN       NaN      NaN      NaN      NaN         NaN  
4    NaN   NaN     NaN       NaN      NaN      NaN      NaN         NaN  
..   ...   ...     ...       ...      ...      ...      ...         ...  
350  NaN   NaN     NaN       NaN      NaN      NaN      NaN         NaN  
351  NaN   NaN     NaN       NaN      NaN

In [15]:
sim.client.showq()

{"asctime": "2023-02-11 21:38:58,037", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Processed command 47", "cmnd_config": {"id": 47, "cmd": "showq -U clos21", "ts": "2023-02-11T21:38:58.030737", "status": "COMPLETE", "stdout": "Using system default configuration file \n\nSUMMARY OF JOBS FOR USER: <clos21>\n\nACTIVE JOBS--------------------\nJOBID     JOBNAME    USERNAME      STATE   NODES REMAINING STARTTIME\n=====================", "stderr": "", "history": [{"ts": "2023-02-11T21:38:56.695536", "status": "STARTED"}], "rt": 1.0, "rc": 0}}
{"asctime": "2023-02-11 21:38:58,037", "name": "taccjm.TACCClient", "levelname": "INFO", "message": "Processed command 47", "cmnd_config": {"id": 47, "cmd": "showq -U clos21", "ts": "2023-02-11T21:38:58.030737", "status": "COMPLETE", "stdout": "Using system default configuration file \n\nSUMMARY OF JOBS FOR USER: <clos21>\n\nACTIVE JOBS--------------------\nJOBID     JOBNAME    USERNAME      STATE   NODES REMAINING STARTTIME\n=========

[{'job_id': '694726',
  'job_name': 'test_20230',
  'username': 'clos21',
  'state': 'Complte',
  'nodes': '1',
  'remaining': '1',
  'start_time': 'Unknown'}]